sdk/storage/Azure.Storage.Common/tests/Shared/RepeatingStream.cs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Buffers;
+using System.IO;
+using Azure.Core;
+
+namespace Azure.Storage.Tests.Shared
+{
+    /// <summary>
+    /// Stream with a defined buffer and length that loops over the buffer for
+    /// a given read at any position. The byte at stream position n is
+    /// buffer[n % buffer.Length].
+    /// </summary>
+    internal class RepeatingStream : Stream
+    {
+        private readonly ReadOnlyMemory<byte> _data;
+        private readonly bool _revealsLength;
+        private readonly long _length;
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => _revealsLength;
+
+        public override bool CanWrite => false;
+        public override long Length => _revealsLength
+            ? _length
+            : throw new NotSupportedException();
+
+        public override long Position { get; set; } = 0;
+
+        public RepeatingStream(int bufferSize, long streamLength, bool revealsLength)
+        {
+            var buf = new byte[bufferSize];
+            new Random().NextBytes(buf);
+            _data = new ReadOnlyMemory<byte>(buf);
```

Comment:
For my own education: Does `ReadOnlyMemory` copy the data here or just somehow keeps `buf` alive (in-scope)?

Reply:
`Memory` and `Span` (and their readonly versions) are facades around existing memory allocations. They're useful for not only being able to specify a parameter as unwritable but also they allow splicing in-place. No byte copies.

Reply:
Makes sense, thanks. But `buf` is going out of scope after this constructor and would then be garbage collected so I guess `ReadOnlyMemory` takes ownership of the memory or something like that (so it doesn't get garbage collected)?

Reply:
I promise it all works out, the same way a field in a class maintains reference to keep alive from the garbage collector.